### ISSUE TITLE
Why is video playback speed delayed when changing and not instant?

### ISSUE THREAD

**Issue by blonsky95** [google/ExoPlayer#7982]
Hi there,

I only post this question here because I have had no luck in Stack Overflow and I've read everything I've been able to found in Exoplayer docs.

I'm using the Exoplayer PlaybackParameters in my app to change the speed of the video playback. However the speed change is only carried out a considerable amount of frames later (I guess it waits for closest sync point?). 

I am aware that when I set the playback parameters it should trigger `default void onPlaybackParametersChanged(PlaybackParameters playbackParameters)
` so I'm not sure if the delay lies in (setting playbackParameters -> triggering the listener) or in (listener triggered -> applies parameters to playback).

If I make a speed change with the video paused, and seek to a different part of the video, the speed change is carried out immediately, it's only while the playback is running that it carries out the speed change with a delay.

Wondering if there is something like SeekParameters where you can choose to seek to closest Sync point, or ignore sync points at all and use tolerance of 0, so setting the PlaybackParameters don't have to wait to the sync point to change.

I couldn't find anything remotely similar in issues, but please let me know if I am missing something out or any existing solutions to my problem

**Comment by Samrobbo** [comment]
Hi, could you try calling [setEnableAudioTrackPlaybackParams(true)](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/DefaultRenderersFactory.html#setEnableAudioTrackPlaybackParams-boolean-) on your `DefaultRenderersFactory` and see if that makes a difference. 

If you don't already have `DefaultRenderersFactory`, initialize and pass it into the constructor of the builder when creating your `SimpleExoPlayer`.

Note that this will only work from API 23 onwards.

For pre-API 23, you could look into a smaller AudioTrack buffer, which would be more code than this helper method, and it may increase the chance of an audio underrun.

**Comment by blonsky95** [comment]
Hi Sam thanks for helping me out,

I just did that and I've been playing around with it for a bit, and it does make it work much quicker, however it doesnt make a smooth transition from the old speed to the new one, and its a bit more clunky:

e.g. if I go speed 1.0 to speed 0.3 in one step, the frames freeze for a second and then start playing at 0.3, and everytime you seek a point in the video (while keeping the playback params speed at 0.3) it does a similar thing, it freezes or starts off at speed 1.0 before adjusting to 0.3

Not sure how helpful that is, but the media audio is quite unsynced (which is not a big issue, just a detail), to the point where when the speed is below 1, e.g. 0.3, the audio goes on after pausing the video. I've tried different frame rates videos and same thing, I might try playing around with LoadControls to change buffering parameters, would it be worth a try?

Without the setEnableAudioTrackPlaybackParams(true) there was a delay but after that, the speed was kept constant, and it didnt feel clunky when seeking throughout video, it was just the delay that was problematic, so something with the best of both of things would be great.

**Comment by Samrobbo** [comment]
@andrewlewis You worked on the speed adjustment stuff, do you know why that might be happening?

**Comment by blonsky95** [comment]
I know it's not the best description, but the set up is fairly simple. I'm initialising with just:
`val defaultRenderersFactory = DefaultRenderersFactory(this).setEnableAudioTrackPlaybackParams(true)`
`exoPlayer = SimpleExoPlayer.Builder(this, defaultRenderersFactory).build()`

and then setting the playback params (speed 0.2 for example) with:
`        exoPlayer.setPlaybackParameters(PlaybackParameters(0.2f))
`

With `setEnableAudioTrackPlaybackParams(true)` , I get the clunky and inconsistent speed changes, but the almost instant speed change. If I seek to different points in the video while speed is low (e.g. 0.2), some frames freeze for a second, or even accelerate before setting the right speed.

When I initialise exoPlayer without `setEnableAudioTrackPlaybackParams(true)`  I get the playback speed change delay, but consistent new speed after that, I can seek different points and I always see 0.2 speed and no freezing.

Let me know if you need any more info

**Comment by andrewlewis** [comment]
Thanks for reporting this. The problem seems to be caused by the logic for interpolating between positions when using `getPlaybackHeadPosition` vs `AudioTimestamp` not taking into account the `AudioTrack` playback params. There is a fix in review and this issue will be updated when it's on the development branch.

**Comment by blonsky95** [comment]
Thanks for fixing it! 
So just to be clear, this will fix the clunkiness when `setEnableAudioTrackPlaybacksParams(true) `, not the delay when setting playback speed right? 

And how can I know when the fix is pulled into development branch? 

And lastly, I'm adding the exoplayer dependency through gradle, is there a release version number for the development branch?

**Comment by andrewlewis** [comment]
> So just to be clear, this will fix the clunkiness when `setEnableAudioTrackPlaybacksParams(true)` , not the delay when setting playback speed right?

Correct. We looked at options to address the delay a while ago but couldn't find a clean/general/easy to implement approach (there is no API we can use to process the audio just before the mixer, so we have to do it upstream of the audio track buffer, which introduces latency).

> And how can I know when the fix is pulled into development branch?

See commit referenced just above. If you could try this out and let us know there are any problems that would be great. I will reopen the issue if so.

> And lastly, I'm adding the exoplayer dependency through gradle, is there a release version number for the development branch?

I'm afraid not. #3859 tracks this. There are instructions [here](https://github.com/google/ExoPlayer#locally) for using a local dependency.

**Comment by blonsky95** [comment]
Unfortunately I'm still having issues, I checked I was in the right branch with the updated file and that's fine, I can see the changes you did in my local files so not sure if it addressed my problem directly, however this time I used a sample video so you can see, I played it with the 2 modes, AudioTrackPlaybackParams set to true (no delay but clunky) and to false (delay but no clunky)

Btw, just to add a bit of context, I'm using this to analyse videos in slow motion, hence the need for some sort of consistent video speed changes, specially at lower speeds closer to 0 (minimum is 0.1 = 10% of normal speed)

Video 1 - setEnableAudioTrackPlaybackParams(false)
https://drive.google.com/file/d/1GdZEz8p3Uze8oreHUoVbU-K42QqNTKYQ/view?usp=sharing
As you can see, when I make a speed change (The slider you see at top controls the speed), there is a delay before the change happens. But every time I press pause and play, it plays smoothly at the set speed. The delay is the only issue.

Video 2 -  setEnableAudioTrackPlaybackParams(true)
https://drive.google.com/file/d/1GXSC6OonBfJeocpKchmDGFZgQ9MXKeGW/view?usp=sharing
In this video, I change the speed to 0.1 at the start, and you can see that by doing this the frame freezes, without pause beign pressed. After, I keep the video speed at 0.1 to the end, but every time I hit pause and play again you can see the speed go up for a few frames and then a frame freeze for a bit (which refers to the clunkiness), and then it continues at 0.1. In this same video I also seek forward a bit and the same thing happens, it's like every time it buffers it doesn't have the right params or something, I'm not sure, but let me know if this helps in any way.

**Comment by andrewlewis** [comment]
Regarding video 1, there isn't much we can do here I'm afraid.

Regarding video 2, please could you try disabling audio entirely and seeing if you see the same behavior? I think it may be caused by frame release timing not taking into account the playback speed in which case it would be reproducible without audio being enabled.

**Comment by blonsky95** [comment]
Yeah regarding video 1, sorry, I meant to use it as a reference of what I'm comparing it to, I understand the problem with the delay and the available APIs, no worries.

> Regarding video 2, please could you try disabling audio entirely

How can I disable audio entirely?

**Comment by andrewlewis** [comment]
> How can I disable audio entirely?

I'd just override `DefaultRenderersFactory.buildAudioRenderers` to make it do nothing, or if you know the index of the audio renderer (you can read it off `EventLogger` output for example) you can call `setRendererDisabled` on the track selector parameters builder.

**Comment by blonsky95** [comment]
It worked!

Did just that, subclassed DefaultRendererFactory and overrode buildAudioRenderers to do nothing, and no, there's no audio, but it definitely works great, 0 clunkiness.

So what can be done to get audio back without messing that up? Is it possible?

**Comment by blonsky95** [comment]
Would that commit fix the audio? make it compatible with the speed params adjusting?

**Comment by andrewlewis** [comment]
I think that's just the same change as the one above going into a release.

Still need to investigate further what's going on (if removing the audio renderer fixed the problem then I think it's not to do with frame release timing) but I'm afraid it will be a while until I get round to it.

Hopefully just turning off audio for low speeds is fine for the time being? To turn audio on and off without recreating the player, you could switch to disabling/reenabling the renderer via track selection parameters as mentioned in my previous comment.

**Comment by zoomGitS** [comment]
Any new suggestion about this problem?

**Comment by blonsky95** [comment]
@zoomGitS Not sure if you need audio or not, but if you dont, the speed changes work without clunkiness or delay, by setting `setEnableAudioTrackPlaybackParams(true)` and this other change which is explained better above by Andrew:

 
> Did just that, subclassed DefaultRendererFactory and overrode buildAudioRenderers to do nothing, and no, there's no audio, but it definitely works great, 0 clunkiness.
>

**Comment by blonsky95** [comment]
Update to this issue I opened almost 5 years ago:

I have revisited this and can confirm that by now adding a DefaultRendererFactory with the setEnableAudioTrackPlaybackParams(true) will fix the delay:

E.g.:
```
val defaultRenderersFactory = DefaultRenderersFactory(context).setEnableAudioTrackPlaybackParams(true)
            exoPlayer = ExoPlayer.Builder(context, defaultRenderersFactory).build()
```

No need to remove audio. @andrewlewis @Samrobbo I'll close this issue, thanks for the help!

**Comment by zhoujielun111** [comment]
谢谢啦！邮件我已经收到了！！
